resolver/src/hosts.rs

```diff
@@ -0,0 +1,121 @@
+use std::collections::HashMap;
+use std::fs;
+use std::time::{Duration, SystemTime};
+use std::io::{BufRead, BufReader};
+use std::fs::File;
+use std::net::{IpAddr, Ipv4Addr, Ipv6Addr};
+use std::sync::Mutex;
+
+lazy_static! {
+    static ref CACHE_MAX_AGE: Duration = Duration::new(5, 0);
+    static ref HOSTS: Mutex<Hosts> = Mutex::new(Hosts::new());
+}
+
+#[derive(Debug)]
+pub struct Hosts {
+    by_name: HashMap<String, Vec<IpAddr>>,
+    expire: SystemTime,
+    path: String,
+    mtime: SystemTime,
+    size: u64,
+}
+
+impl Hosts {
+    pub fn new() -> Hosts {
+        let mut hosts = Hosts {
+            by_name: HashMap::new(),
+            expire: SystemTime::now() + *CACHE_MAX_AGE,
+            path: String::from("/etc/hosts"),
+            mtime: SystemTime::now(),
+            size: 0,
+        };
+
+        hosts.update();
+        hosts
+    }
+
+    fn update(&mut self) {
+        let now = SystemTime::now();
+        if now < self.expire && self.by_name.len() > 0 {
+            return;
+        }
+
+        let meta = if let Ok(meta) = fs::metadata(&self.path) {
+            meta
+        } else {
+            return;
+        };
+
+        if self.mtime == meta.modified().unwrap_or(SystemTime::now()) && self.size == meta.len() {
+            self.expire = now + *CACHE_MAX_AGE;
+            return;
+        }
+
+        let f = if let Ok(f) = File::open(&self.path) {
+            f
+        } else {
+            return;
+        };
+
+        for line in BufReader::new(f).lines() {
+            let line = line.unwrap_or_default();
+            let line = if let Some(pos) = line.find('#') {
+                String::from(line.split_at(pos).0)
+            } else {
+                line
+            };
+            let line = line.trim();
+            if line.is_empty() {
+                continue;
+            }
+
+            let fields: Vec<String> = line.split_whitespace().map(|s| s.to_string()).collect();
+            if fields.len() < 2 {
+                continue;
+            }
+            let addr = if let Some(a) = parse_literal_ip(&fields[0]) {
+                a
+            } else {
+                continue;
+            };
+
+            for i in 1..fields.len() {
+                let domain = fields[i].to_lowercase();
+                self.by_name.entry(domain).or_insert(vec![]).push(addr);
+            }
+        }
+
+        self.mtime = meta.modified().unwrap_or(SystemTime::now());
+        self.size = meta.len();
+        self.expire = SystemTime::now() + *CACHE_MAX_AGE;
+    }
+}
+
+// lookup_static_host looks up the addresses for the given host from /etc/hosts.
+pub fn lookup_static_host(host: &str) -> Option<Vec<IpAddr>> {
+    let mut hosts = HOSTS.lock().unwrap();
```

Comment:
straight locks of this data structure will cause potential issues in Tokio. I am thinking it would be easier to use ThreadLocals for this HOSTS cache.